src/LanguageServer/Protocol/Workspaces/ILspMiscellaneousFilesWorkspace.cs

```diff
@@ -0,0 +1,15 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+internal interface ILspMiscellaneousFilesWorkspace : ILspWorkspace, ILspService
+{
+    TextDocument? AddMiscellaneousDocument(Uri uri, SourceText documentText, string languageId, ILspLogger logger);
+    void TryRemoveMiscellaneousDocument(Uri uri, bool removeFromMetadataWorkspace);
+}
```

Comment:
up to you, but i wouldn't mind docs.

Also, being an ILspWorkspace *and* an ILspService seems odd.  I don't really understand the role this plays as both those types.

Reply:
also, as an interface, do we expect multiple impls of this?  i'd love an explanation of that.  thanks!

Reply:
I think the misc workspace *is* an ILspService currently, that's why LspWorkspaceManagerFactory can call `GetService<LspMiscellaneousFilesWorkspace>()`.

The plan is for FileBasedProgramWorkspace to *also* implement the new interface. Most likely we are going to have an opt-out switch for file-based programs support, which simply makes it so the "existing" misc files workspace gets used instead of the "new" FileBasedProgramWorkspace.